### ISSUE TITLE
Issues on develop when updating indexes during "protected-publish"

### ISSUE THREAD

**Issue by alalazo** [LLNL/spack#38772]
### Steps to reproduce

Reported by @haampie 

https://gitlab.spack.io/spack/spack/-/jobs/7594473

We need to investigate and find a reproducer for these warnings

### Error message
```console
$ spack buildcache update-index --keys "${SPACK_COPY_BUILDCACHE}"
==> Warning: cannot reconstruct virtual dependencies on package py-scipy: cannot load package 'py-scipy' from the 'builtin' repository: maintainers() got an unexpected keyword argument 'when'
==> Warning: cannot reconstruct virtual dependencies on package meson: cannot load package 'meson' from the 'builtin' repository: maintainers() got an unexpected keyword argument 'when'
==> Warning: cannot reconstruct virtual dependencies on package umpire: cannot load package 'umpire' from the 'builtin' repository: maintainers() got an unexpected keyword argument 'when'
==> Warning: cannot reconstruct virtual dependencies on package vtk-h: cannot load package 'vtk-h' from the 'builtin' repository: maintainers() got an unexpected keyword argument 'when'
==> Warning: cannot reconstruct virtual dependencies on package parmetis: cannot load package 'parmetis' from the 'builtin' repository: directive "version" cannot be used within a "when" context since it does not support a "when=" argument
==> Warning: cannot reconstruct virtual dependencies on package metis: cannot load package 'metis' from the 'builtin' repository: maintainers() got an unexpected keyword argument 'when'
==> Warning: cannot reconstruct virtual dependencies on package gnuconfig: cannot load package 'gnuconfig' from the 'builtin' repository: maintainers() got an unexpected keyword argument 'when'
==> Warning: cannot reconstruct virtual dependencies on package adiak: cannot load package 'adiak' from the 'builtin' repository: maintainers() got an unexpected keyword argument 'when'
==> Warning: cannot reconstruct virtual dependencies on package protobuf: cannot load package 'protobuf' from the 'builtin' repository: maintainers() got an unexpected keyword argument 'when'
==> Warning: cannot reconstruct virtual dependencies on package libtiff: cannot load package 'libtiff' from the 'builtin' repository: maintainers() got an unexpected keyword argument 'when'
==> Warning: cannot reconstruct virtual dependencies on package pmix: cannot load package 'abduco' from the 'builtin' repository: directive "version" cannot be used within a "when" context since it does not support a "when=" argument
==> Warning: cannot reconstruct virtual dependencies on package py-smart-open: cannot load package 'py-smart-open' from the 'builtin' repository: maintainers() got an unexpected keyword argument 'when'
==> Warning: cannot reconstruct virtual dependencies on package py-pandas: cannot load package 'py-pandas' from the 'builtin' repository: maintainers() got an unexpected keyword argument 'when'
==> Warning: cannot reconstruct virtual dependencies on package py-cppy: cannot load package 'py-cppy' from the 'builtin' repository: maintainers() got an unexpected keyword argument 'when'
==> Warning: cannot reconstruct virtual dependencies on package perl-html-tagset: cannot load package 'perl-html-tagset' from the 'builtin' repository: directive "version" cannot be used within a "when" context since it does not support a "when=" argument
==> Warning: cannot reconstruct virtual dependencies on package py-pydeprecate: cannot load package 'py-pydeprecate' from the 'builtin' repository: directive "version" cannot be used within a "when" context since it does not support a "when=" argument
==> Warning: cannot reconstruct virtual dependencies on package py-deepdiff: cannot load package 'py-deepdiff' from the 'builtin' repository: directive "version" cannot be used within a "when" context since it does not support a "when=" argument
==> Warning: cannot reconstruct virtual dependencies on package py-dm-tree: cannot load package 'py-dm-tree' from the 'builtin' repository: maintainers() got an unexpected keyword argument 'when'
==> Warning: cannot reconstruct virtual dependencies on package perl-xml-parser: cannot load package 'perl-xml-parser' from the 'builtin' repository: directive "version" cannot be used within a "when" context since it does not support a "when=" argument
==> Warning: cannot reconstruct virtual dependencies on package lerc: cannot load package 'lerc' from the 'builtin' repository: directive "version" cannot be used within a "when" context since it does not support a "when=" argument
==> Warning: cannot reconstruct virtual dependencies on package perl-module-build: cannot load package 'perl-module-build' from the 'builtin' repository: directive "version" cannot be used within a "when" context since it does not support a "when=" argument
==> Warning: cannot reconstruct virtual dependencies on package py-numexpr: cannot load package 'py-numexpr' from the 'builtin' repository: directive "version" cannot be used within a "when" context since it does not support a "when=" argument
==> Warning: cannot reconstruct virtual dependencies on package py-stopit: cannot load package 'py-stopit' from the 'builtin' repository: directive "version" cannot be used within a "when" context since it does not support a "when=" argument
==> Warning: cannot reconstruct virtual dependencies on package perl-http-date: cannot load package 'perl-http-date' from the 'builtin' repository: directive "version" cannot be used within a "when" context since it does not support a "when=" argument
==> Warning: cannot reconstruct virtual dependencies on package py-importlib-resources: cannot load package 'py-importlib-resources' from the 'builtin' repository: directive "version" cannot be used within a "when" context since it does not support a "when=" argument
==> Warning: cannot reconstruct virtual dependencies on package py-ordered-set: cannot load package 'py-ordered-set' from the 'builtin' repository: directive "version" cannot be used within a "when" context since it does not support a "when=" argument
==> Warning: cannot reconstruct virtual dependencies on package llvm-openmp-ompt: cannot load package 'llvm-openmp-ompt' from the 'builtin' repository: directive "version" cannot be used within a "when" context since it does not support a "when=" argument
Cleaning up project directory and file based variables
```
### Information on your system

runner-x86-v3-pub-gitlab-runner-d6c474968-kplgk

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by alalazo** [comment]
Must be related to #34821

**Comment by haampie** [comment]
I think it's more likely that something throws while a directive is being evaluated, something like that happened to me in the past

**Comment by haampie** [comment]
Closed by #39594
